gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcher.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * The user of this GoogleWebmasterDataFetcher should follow these steps:
+ * 1. Call getAllPages to get all pages based on your filters.
+ * 2. For each page we get at last step, call performSearchAnalyticsQuery to get query data. If query data is not available, then this page won't be included in the return value.
+ *
+ * Note:
+ * Due to the limitation of the Google API -- each API request can return at most 5000 rows of records, you need to take additional steps to fetch all data that the Google Search Console keeps.
+ *
+ * There is a rule to check whether Google Search Console keeps more than 5000 rows of data based on your request. The rule is that if you request for 5000 rows, but the response has less than 5000 rows; then in this case you've got all data that the Google service provides. If you request for 5000 rows, and the API returns you 5000 rows, there is a high chance that the Google service has more data matching your query, but due to the limitation of the API, only first 5000 rows returned.
+ *
+ */
+public abstract class GoogleWebmasterDataFetcher {
+  public abstract String getSiteProperty();
+
+  enum Metric {
+    CLICKS, IMPRESSIONS, CTR, POSITION
+  }
+
+  /**
+   * Results are composed of [[requestedDimension list], clicks, impressions, ctr, position]
+   * @param rowLimit row limit for this API call
+   * @param requestedDimensions a list of dimension requests. The dimension values can be found at the first part of the return value
+   * @param filters filters of your request
+   */
+  public abstract List<String[]> performSearchAnalyticsQuery(String startDate, String endDate, int rowLimit,
+      List<GoogleWebmasterFilter.Dimension> requestedDimensions, List<Metric> requestedMetrics, Collection<ApiDimensionFilter> filters)
+      throws IOException;
+
+  /**
+   * Call API in batches
+   */
+  public abstract void performSearchAnalyticsQueryInBatch(List<ProducerJob> jobs,
+      List<ArrayList<ApiDimensionFilter>> filterList,
+      List<JsonBatchCallback<SearchAnalyticsQueryResponse>> callbackList, List<GoogleWebmasterFilter.Dimension> requestedDimensions,
+      int rowLimit) throws IOException;
+
+  /**
+   * Return all pages given (date, country) filter
+   * @param country country code string
+   * @param rowLimit this is mostly for testing purpose. In order to get all pages, set this to the API row limit, which is 5000
+   */
+  public abstract Collection<ProducerJob> getAllPages(String startDate, String endDate, String country, int rowLimit)
+      throws IOException;
+
+  public static List<String[]> convertResponse(List<Metric> requestedMetrics, SearchAnalyticsQueryResponse response) {
+    List<ApiDataRow> rows = response.getRows();
+    if (rows == null || rows.isEmpty()) {
+      return new ArrayList<>();
+    }
+    List<String[]> ret = new ArrayList<>(rows.size());
+    for (ApiDataRow row : rows) {
+      List<String> keys = row.getKeys();
+      String[] data = new String[keys.size() + 4];
```

Comment:
requestedMetrics might be lesser or more than 4 now / or in future. So, initializing data[] with keys.size() + 4 might run into an issue. how about either making it dynamic or applying a Precondition and documenting this